File: c/problem_8.c
#include "problem_8.h"
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

INode *newINode(int val, INode *left, INode *right) {
    INode *n = malloc(sizeof(INode));
    n->val = val; n->left = left; n->right = right;
    return n;
}

void freeITree(INode *root) {
    if (!root) return;
    freeITree(root->left); freeITree(root->right); free(root);
}

/* helper returns is_unival via out-param; returns the subtree count. */
static int helper(INode *node, int *isUnival) {
    if (!node) { *isUnival = 1; return 0; }
    int lv, rv;
    int lc = helper(node->left,  &lv);
    int rc = helper(node->right, &rv);
    int count = lc + rc;
    *isUnival = lv && rv
        && (!node->left  || node->left->val  == node->val)
        && (!node->right || node->right->val == node->val);
    if (*isUnival) count++;
    return count;
}

int countUnival(INode *root) {
    int dummy;
    return helper(root, &dummy);
}

void problem8(void) {
    printf("Problem 8:\n");
    INode *root = newINode(0,
        newINode(1, NULL, NULL),
        newINode(0,
            newINode(1, newINode(1,NULL,NULL), newINode(1,NULL,NULL)),
            newINode(0, NULL, NULL)
        )
    );
    assert(countUnival(root) == 5);
    freeITree(root);
    printf("All Tests Passed\n");
}

File: rust/problem_8.rs
// IntNode is a binary tree node with an integer value.
pub struct IntNode {
    pub val: i64,
    pub left: Option<Box<IntNode>>,
    pub right: Option<Box<IntNode>>,
}

impl IntNode {
    pub fn new(val: i64, left: Option<Box<IntNode>>, right: Option<Box<IntNode>>) -> Box<Self> {
        Box::new(IntNode { val, left, right })
    }
    pub fn leaf(val: i64) -> Box<Self> { Self::new(val, None, None) }
}

// count_unival_subtrees counts subtrees where every node has the same value.
// Returns (is_unival, count).
fn helper(node: &Option<Box<IntNode>>) -> (bool, usize) {
    let node = match node { Some(n) => n, None => return (true, 0) };
    let (left_unival,  left_count)  = helper(&node.left);
    let (right_unival, right_count) = helper(&node.right);
    let mut count = left_count + right_count;

    let is_unival = left_unival && right_unival
        && node.left.as_ref().map_or(true,  |n| n.val == node.val)
        && node.right.as_ref().map_or(true, |n| n.val == node.val);

    if is_unival { count += 1; }
    (is_unival, count)
}

pub fn count_unival_subtrees(root: &Option<Box<IntNode>>) -> usize {
    helper(root).1
}

pub fn problem8() {
    println!("Problem 8:");
    //       0
    //      / \
    //     1   0
    //        / \
    //       1   0
    //      / \
    //     1   1
    let root = IntNode::new(0,
        Some(IntNode::leaf(1)),
        Some(IntNode::new(0,
            Some(IntNode::new(1, Some(IntNode::leaf(1)), Some(IntNode::leaf(1)))),
            Some(IntNode::leaf(0)),
        )),
    );
    assert_eq!(count_unival_subtrees(&Some(root)), 5, "assertion failed");
    println!("All Tests Passed");
}
